FreeRTOS_IP.c

```diff
@@ -3097,7 +3097,24 @@ static eFrameProcessingResult_t prvProcessIPPacket( IPPacket_t * pxIPPacket,
                     else
                 #endif /* ipconfigUSE_IPv6 */
                 {
-                    vARPRefreshCacheEntry( &( pxIPPacket->xEthernetHeader.xSourceAddress ), pxIPHeader->ulSourceIPAddress, pxNetworkBuffer->pxEndPoint );
+                    IPV4Parameters_t * pxIPv4Settings = &( pxNetworkBuffer->pxEndPoint->ipv4_settings );
+
+                    if( ( pxIPHeader->ulSourceIPAddress & pxIPv4Settings->ulNetMask ) == ( pxIPv4Settings->ulIPAddress & pxIPv4Settings->ulNetMask ) )
+                    {
+                        /* If the IP is on the same subnet and we do not have an ARP entry already,
+                         * then we should send out ARP for finding the MAC address. */
+                        if( xIsIPInARPCache( pxIPHeader->ulSourceIPAddress ) == pdFALSE )
+                        {
+                            FreeRTOS_OutputARPRequest( pxIPHeader->ulSourceIPAddress );
+                            /* Drop this packet. */
+                            eReturn = eReleaseBuffer;
```

Comment:
Here either a TCP or an ICMP packet was received. It will be dropped because the sender's IPv4 address does not occur in the ARP cache yet. Which is probably good and correct. I would like to test if it will happen often that packets are dropped here, and how it affects existing projects.

Note that every entry in the ARP cache is checked regularly by the function `vARPAgeCache()`. Entries that are about to expire will send an ARP request.